src/source/tile_id.js

```diff
@@ -164,11 +164,26 @@ export class OverscaledTileID {
     }
 }
 
-function calculateKey(wrap: number, z: number, x: number, y: number) {
+function calculateKey(wrap: number, overscaledZ: number, z: number, x: number, y: number): string {
     wrap *= 2;
     if (wrap < 0) wrap = wrap * -1 - 1;
     const dim = 1 << z;
-    return ((dim * dim * wrap + dim * y + x) * 32) + z;
+    let id = `${dim * dim * wrap + dim * y + x}`;
+    // Always reserve two characters for both z values in order to
+    // keep the sorting behavior consistent
+    if (z < 10)
+        id += "0";
+    id += z;
+    if (overscaledZ < 10)
+        id += "0";
+    id += overscaledZ;
```

Comment:
We could use non-base 10 number to string conversions, e.g. `overscaledZ.toString(36)`, same for the other numbers. Would be interesting to see if shorter strings can be interned better, or are stored directly inline in the JSValue node.

Reply:
This is a very good point. I tried to keep functionality and logic in `calculateKey` as simple as possible because I noticed that `LayerRaster` benchmark is very sensitive to any changes in the key calculation function. These findings have resulted in some really good optimizations in `findLoadedParent` and `draw_raster` functions by @astojilj so they're not bottlenecks anymore. Non-base 10 encodings would definitely be better so I'll see what performance implications there might be.